examples/notification-demo/echo_server/echo.c

```diff
@@ -0,0 +1,1975 @@
+    struct ca_cert *issuer_cert;
+    int error_code;
+};
+
+
+static int myVerifyCheck(int preverify, WOLFSSL_X509_STORE_CTX* store) {
+    struct echo_server_data *app_data = NULL;
+    const char* issuerName = NULL;
+    unsigned long issuerHash = 0;
+#if defined(DEBUG_TLS) && defined(OPENSSL_EXTRA)
+    const char *subjectCN;
+    const char* subjectName;
+    unsigned long subjectHash = 0;
+#endif
+
+    (void)preverify; /* incoming validation of cert 1=okay, 0=failed */
+
+    if (store == NULL) {
+        fprintf(stderr,"%s L%d null store!\n",__FILE__,__LINE__);
+        return WOLFSSL_FAILURE;
+    }
+
+    app_data = (struct echo_server_data *)store->userCtx;
+    if (app_data == NULL) {
+        fprintf(stderr,"%s L%d null store->userCtx!\n",__FILE__,__LINE__);
+        return WOLFSSL_FAILURE;
+    }
+
+    app_data->error_code = NO_PEER_CERT; /* default error code */
+
+    if (store->current_cert == NULL) {
+        fprintf(stderr,"%s L%d null store->current_cert\n",__FILE__,__LINE__);
+        return WOLFSSL_FAILURE;
+    }
+
+    if (store->error != 0) {
+#ifdef DEBUG_TLS
+        fprintf(stderr, "%s L%d depth=%d ->error=%d (%s)\n",__FILE__,__LINE__,
+            store->error_depth,store->error,
+            wolfSSL_ERR_reason_error_string(store->error));
+#endif
+        app_data->error_code = store->error;
+        return WOLFSSL_FAILURE;
+    }
+
+    if (store->error_depth != 0) /* "0=peer, >1 intermediates" */ {
+        app_data->error_code = 0;
+        return WOLFSSL_SUCCESS;
+    }
+
+#ifdef OPENSSL_EXTRA
+    issuerHash = wolfSSL_X509_issuer_name_hash(store->current_cert);
+    issuerName  = wolfSSL_X509_NAME_oneline(
+            wolfSSL_X509_get_issuer_name(store->current_cert), 0, 0);
+
+    #ifdef DEBUG_TLS
+    subjectCN = wolfSSL_X509_get_subjectCN(store->current_cert);
+    subjectHash = wolfSSL_X509_subject_name_hash(store->current_cert);
+    subjectName = wolfSSL_X509_NAME_oneline(
+            wolfSSL_X509_get_subject_name(store->current_cert), 0, 0);
+    printf("CN=%s subjectName->name=\"%s\" subject_hash=%lu\n",
+        subjectCN, subjectName, subjectHash);
+    printf("issuerName->name=\"%s\" issuer_hash=%lu\n", issuerName, issuerHash);
+    #endif
+#endif
+
+    for (struct ca_cert *ca_certs_i = ca_certs;
+         ca_certs_i < &ca_certs[sizeof ca_certs / sizeof ca_certs[0]];
+         ++ca_certs_i)
+    {
+        if ((issuerHash == ca_certs_i->subjectHash) &&
+                (!strncmp(issuerName, ca_certs_i->subject, WC_ASN_NAME_MAX))) {
+            app_data->issuer_cert = ca_certs_i;
+            break;
+        }
+    }
+    if (app_data->issuer_cert == NULL) {
+        fprintf(stderr, "%s L%d app_data->issuer_cert NULL.\n",__FILE__,__LINE__);
+        return WOLFSSL_FAILURE;
+    }
+
+    app_data->error_code = 0;
+
+    return WOLFSSL_SUCCESS;
+}
+
+/* matches logic in wolfSSL_X509_issuer_name_hash() and MakeWordFromHash(). */
+static unsigned long cert_subject_name_hash(DecodedCert *cert) {
```

Comment:
I wonder if we really need to nix this routine, the related `ca_cert.subjectHash`, the call to it in `echo_ssl_init()`, and the related comparison above, and just use the `strncmp()`.  `cert_subject_name_hash()` seems like a pretty gratuitous abstraction barrier violation, and quite fidgity.

Reply:
It is fine for now. Not changing.